waspls/src/Wasp/Backend/ConcreteParser.hs

```diff
@@ -0,0 +1,98 @@
+module Wasp.Backend.ConcreteParser
+  ( -- * Concrete parser
+
+    -- | This module contains functions for converting lists of "Token"s into
+    -- concrete syntax trees. These trees represent any source file completely,
+    -- including whitespace and comments, even if the source file is invalid.
+    parseCST,
+    parseCSTExpression,
+  )
+where
+
+import Wasp.Backend.ConcreteParser.Internal
+import Wasp.Backend.ConcreteSyntax (SyntaxKind (..), SyntaxNode)
+import Wasp.Backend.ParseError
+import Wasp.Backend.Token (Token)
+import qualified Wasp.Backend.Token as T
+
+-- | Parse a list of tokens into a concrete syntax tree for an entire wasp file.
+--
+-- See "SyntaxNode" for a description of what a concrete syntax tree contains.
+parseCST :: [Token] -> ([ParseError], [SyntaxNode])
+parseCST tokens = parse tokens (root --> eof)
+
+-- | Parse a list of tokens into a concrete syntax tree for a wasp expression.
+-- This is mainly used for testing.
+--
+-- See "SyntaxNode" for a description of what a concrete syntax tree contains.
+parseCSTExpression :: [Token] -> ([ParseError], [SyntaxNode])
+parseCSTExpression tokens = parse tokens (expr --> eof)
+
+root :: GrammarRule
+root = group Program stmts
+
+stmts :: GrammarRule
+stmts = eof <> (stmt --> stmts)
+
+stmt :: GrammarRule
+stmt =
+  group Decl $
+    (T.Identifier `as` DeclType)
+      --> (T.Identifier `as` DeclName)
+      --> expr
```

Comment:
Based on discussion from above, if we are going with `<>` for chaining, then I think we are good! Because we can then do `<>`, but we can also do `mconcat [...]`, so we can format any way we like. So I don't think we need to change anything regarding how `<$$>` operates. I will let you resolve this comment once you read this message of mine.